core/clients/metadataclients/device_test.go

```diff
@@ -14,88 +14,58 @@
 package metadataclients
 
 import (
-	"fmt"
 	"github.com/edgexfoundry/edgex-go/core/domain/models"
-	"gopkg.in/mgo.v2/bson"
-	"os"
+	"net/http"
+	"net/http/httptest"
 	"testing"
 )
 
 const (
-	deviceUrl        = "http://localhost:48081/api/v1/device"
-	addressableUrl   = "http://localhost:48081/api/v1/addressable"
-	deviceServiceUrl = "http://localhost:48081/api/v1/deviceservice"
-	deviceProfileUrl = "http://localhost:48081/api/v1/deviceprofile"
+	DeviceUriPath = "/api/v1/device"
 )
 
 // Test adding a device using the device client
 func TestAddDevice(t *testing.T) {
 	d := models.Device{
-		Addressable:    a,
+		Id:             "1234",
+		Addressable:    models.Addressable{},
 		AdminState:     "UNLOCKED",
 		Name:           "Test name for device",
 		OperatingState: "ENABLED",
-		Profile:        dp,
-		Service:        ds,
+		Profile:        models.DeviceProfile{},
+		Service:        models.DeviceService{},
 	}
 
-	_, err := dc.Add(&d)
-	if err != nil {
-		t.Log(err.Error())
-		t.FailNow()
-	}
-}
+	addingDeviceId := d.Id.Hex()
 
-var dc DeviceClient
-var a models.Addressable
-var ds models.DeviceService
-var dp models.DeviceProfile
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusOK)
 
-// Main method for the tests
-func TestMain(m *testing.M) {
-	dc = NewDeviceClient(deviceUrl)
-	ac := NewAddressableClient(addressableUrl)
-	dsc := NewServiceClient(deviceServiceUrl)
-	dpc := NewDeviceProfileClient(deviceProfileUrl)
+		if r.Method != http.MethodPost {
+			t.Errorf("expected http method is %s, active http method is : %s", http.MethodPost, r.Method)
+		}
 
-	a = models.Addressable{
-		Address: "http://localhost",
-		Name:    "Test Addressable",
-		Port:    3000,
-	}
-	id, err := ac.Add(&a)
-	if err != nil {
-		fmt.Println("Error posting addressable: " + err.Error())
-		return
-	}
-	a.Id = bson.ObjectIdHex(id)
+		if r.URL.EscapedPath() != DeviceUriPath {
+			t.Errorf("expected uri path is %s, actual uri path is %s", DeviceUriPath, r.URL.EscapedPath())
+		}
 
-	ds = models.DeviceService{
-		AdminState: "UNLOCKED",
-		Service: models.Service{
-			Addressable:    a,
-			Name:           "Test device service",
-			OperatingState: "ENABLED",
-		},
-	}
-	id, err = dsc.Add(&ds)
-	if err != nil {
-		fmt.Println("Error posting device service: " + err.Error())
-		return
-	}
-	ds.Service.Id = bson.ObjectIdHex(id)
+		w.Write([]byte(addingDeviceId))
 
-	dp = models.DeviceProfile{
-		Manufacturer: "Test manufacturer for device profile",
-		Model:        "Test model for device profile",
-		Name:         "Test name for device profile",
-	}
-	id, err = dpc.Add(&dp)
+	}))
+
+	defer ts.Close()
+
+	url := ts.URL + DeviceUriPath
+
+	dc := NewDeviceClient(url)
+
+	receivedDeviceId, err := dc.Add(&d)
```

Comment:
Add method should return predictable output:
* it return device_id and err == nil
* return err

In other words it should be mocked and not depend on Add() logic because in such case it integration not unit test
In other words it should be mocked and does not depend on Add() logic because in such case it integration not unit test

Reply:
@falkerson  you mean I should mock every return branch to cover the test coverage? Actually, I tested rest client. Because it is not necessary to test the client itself. I just want test the full http rest flow.

Reply:
@yanghua ideally yes. Unit testing tests piece of codes on some logic with most granular and isolated level and it shouldn't have any external dependencies like: open file descriptors, network connects etc. Also it shouldn't depend on output of some another function because behavior of unit test became unpredictable. For me if we have 'if' statements in unit test means thats smth wrong with test itself and if couldn't easily mock some internal functions smth wrong with code design. And full flow testing is kind of integration test. But it depends on what "unit" means for us (https://martinfowler.com/articles/mocksArentStubs.html#TestIsolation)

Reply:
@falkerson I agree the unit test case should promote the test coverage, it's best practice. But I don't agree the unit test should not contain "if statements", here you can consider it as `assert` `assertEquals` in JUnit (the most popular unit test framework for Java), the different is `assert` encapsulates the "if statements". About the external dependencies, I think it depends on the context of code. Just like here, the rest client must make a http request to the microservice, it's the part of the business logic. Consider, if it can not make a http request and get a response, Can it return successfully? Of cause not. So for test it, I mocked a local http server used Golang's httptest framework and let the `Add` method passthrough normally. 

If there is another method use `DeviceClient` we could implement a `MockDeviceRestClient` and let the `Add` method do not make a real http request (for example, just return a const). But here we need test a rest client itself, so we need mock a http server.

So I think we should promote the test coverage, I will do it if I have enough time. This issue's purpose is remove the dependency to  real microservices.

Reply:
@yanghua about 'if' statement I mean:

receivedDeviceId, err := dc.Add(&d)

So in one case err == nil and in another not. What case we have in this test case function? dc.Add(&d) should be **idempotent** and return same result every time test runs. If err will not be nil test ends with t.Error(err.Error()) and if receivedDeviceId != addingDeviceId doesn't make sense. In another case if err !=nil {} is redundant because err always be nil.

Reply:
@falkerson you are caught in this function's current implementation, `Add` method returns error also may cause by the `json.Marshal`, I said when I have enough time I would give more input data case to promote the test coverage. If I build a invalidated Device model, it will return error.